servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/CompletableMergeWithPublisher.java

```diff
@@ -34,22 +37,152 @@
+                    subscriber, contextMap));
+            completableSubscriber = new CompletableSubscriber();
+        }
+
+        void merge(Completable original, Publisher<? extends T> mergeWith, SignalOffloader signalOffloader,
+                   AsyncContextMap contextMap, AsyncContextProvider contextProvider) {
+            offloadedSubscriber.onSubscribe(
+                    new MergedCancellableWithSubscription(subscription, completableSubscriber));
+            original.delegateSubscribe(completableSubscriber, signalOffloader, contextMap,
+                    contextProvider);
+            // SignalOffloader is associated with the original Completable. Since mergeWith Publisher is provided by
+            // the user, it will have its own Executor, hence we should not pass this signalOffloader to subscribe to
+            // mergeWith.
+            // Any signal originating from mergeWith Publisher should be offloaded before they are sent to the
+            // Subscriber of the resulting Publisher of CompletableMergeWithPublisher as the Executor associated with
+            // the original Completable defines the threading semantics for that Subscriber.
+            mergeWith.subscribeInternal(this);
+        }
+
+        @Override
+        public void onSubscribe(final Subscription subscription) {
+            this.subscription.delayedSubscription(subscription);
+        }
+
+        @Override
+        public void onNext(@Nullable final T t) {
+            offloadedSubscriber.onNext(t);
+        }
+
+        @Override
+        public void onError(final Throwable t) {
+            terminateSubscriber(new TerminalSignal(t, true));
+        }
+
+        @Override
+        public void onComplete() {
+            terminateSubscriber(TerminalSignal.PUB_COMPLETED);
+        }
+
+        private void terminateSubscriber(final TerminalSignal terminalSignal) {
+            for (;;) {
+                final TerminalSignal currState = terminal;
+                if (currState != null) {
+                    if (currState.fromPublisher == terminalSignal.fromPublisher) {
+                        // The goal of this check is to prevent concurrency on the Subscriber and require both sources
+                        // terminate before terminating downstream. We don't need to enforce each source terminates
+                        // exactly once for the correctness of this operator (so we don't).
+                        throw duplicateTerminalException(currState);
+                    }
+                    if (currState.cause == null) {
+                        if (terminalSignal.cause == null) {
+                            offloadedSubscriber.onComplete();
+                        } else {
+                            offloadedSubscriber.onError(terminalSignal.cause);
+                        }
+                    } else {
+                        offloadedSubscriber.onError(currState.cause);
```

Comment:
Looking at `CompletableMergeSubscriber`, there we add the 2nd error as suppressed for the 1st one, but here we propagate only the first failure. The 2nd may also be useful, consider propagating both.

Reply:
I'll add it but I'm hesitant to use `addSuppressed` in general as it may result in memory leak with static exceptions, difficult to "turn off" and limit the queue size ... but we can look at this issue more generally in a followup.

Reply:
actually unit tests remind me of another issue with `addSuppressed`, self-suppression throws an exception so we would have to guard against this everywhere (which we don't). I'll skip `addSuppressed` for now...